#!/bin/bash

#PBS -N step1epi

#PBS -l select=1:ncpus=6:mem=64gb

#PBS -l walltime=11:59:00

#PBS -j oe

#PBS -m ae

#PBS -M INSERT_EMAIL

set -e # exit if any command fails

#############################  EDIT ME  #######################################

# Working directory (where output directory will be created).
my_dir="/srv/scratch/zid"
# Name of output folder and file prefixes.
output="INSERT_FOLDER_NAME_HERE"
# Path to the EpiNano directory
epinano_dir="$my_dir/EpiNano"
# Reference genome path:
genome_path="$my_dir/Homo_sapiens/UCSC/hg38/Sequence/WholeGenomeFasta/genome.fa"
# Basecall reads fastq path:
basecall_path="$my_dir/dir/pass/tissue.fastq.gz"

###############################################################################


module load minimap2/2.17
module load samtools/1.10

cd $my_dir
mkdir -p $output/mapped
cd $output/mapped

# STEP 0. Align basecall reads to genome (not transcriptome).
## 0.1 Mapped output created in '$my_dir/$output/mapped'.
minimap2 --MD -t 6 -ax splice -k14 -uf $genome_path $basecall_path > ${output}_aln.sam

## 0.2 Convert mapping SAM file to BAM
bam_output_name="sorted_${output}.bam"
samtools view -hbS -F 3844 $my_dir/$output/mapped/${output}_aln.sam | samtools sort -@ 6 -o $bam_output_name

# Split the BAM file into smaller BAM files (by chromosome).
bamtools split -in $my_dir/$output/mapped/$bam_output_name -reference

# Index each new BAM file.
ls $my_dir/$output/mapped/sorted_${output}.REF_*.bam | xargs -n1 -P5 samtools index